If we re-fakeify a FakeTensor with the same ShapeEnv, preserve symbols

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113690
* #113080
* #111554
* #113666
* #113653
* __->__ #113651

Subsumes half of https://github.com/pytorch/pytorch/pull/113605

We support fakeifying an already fake tensor, which will give you a new fake tensor mirroring the same structure as the original fake tensor, which is what is needed by https://github.com/pytorch/pytorch/issues/113643 . However, when this refakeification happens, we will naively reallocate all new sizes for all of the fake tensor. This is the right thing to do if you are re-fakeifying on a fresh ShapeEnv (because you're reparametrizing the sizes or something), but if you have two fake tensor modes which are sharing a shape environment, you would actually rather just reuse the original sizes/strides/offset from the original fake tensor. This ends up being pretty simple. I recommend viewing with whitespace diff turned off.

There's some fuzz around jagged tensor handling; that code is probably not quite right, but I fixed it for this particular case in the most straightforward way.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
If we re-fakeify a FakeTensor with the same ShapeEnv, preserve symbols

Subsumes half of https://github.com/pytorch/pytorch/pull/113605

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD reviewed (approved):
nit could have a test where the shape env don't match and ensure new symints are re-generated?

thiagocrepaldi reviewed (commented):
https://github.com/pytorch/pytorch/pull/113605#issuecomment-1810640367

eellison commented on torch/_subclasses/meta_utils.py:
```diff
@@ -474,31 +483,43 @@ def empty_create(inner_t, inner_src):
+                                # be a variable
+                                ctx["ragged_size"] = t._size[1]
+                            else:
+                                assert t._size[1].node.singleton_int() is not None
```
I guess for this to be in the generic mechanism, we would need to change the api of `__tensor_flatten__` ?

ezyang replied:
Yeah, I feel there is something wrong with the code here, but not entirely sure what. I mentioned this to @soulitzer. At least part of the problem is that the ragged special case code doesn't correctly respect DimDynamic

jbschlosser replied:
In #113279, I'm working on correctly respecting DimDynamic

voznesenskym committed:
Update on "If we re-fakeify a FakeTensor with the same ShapeEnv, preserve symbols"


Subsumes half of https://github.com/pytorch/pytorch/pull/113605

We support fakeifying an already fake tensor, which will give you a new fake tensor mirroring the same structure as the original fake tensor, which is what is needed by https://github.com/pytorch/pytorch/issues/113643 . However, when this refakeification happens, we will naively reallocate all new sizes for all of the fake tensor. This is the right thing to do if you are re-fakeifying on a fresh ShapeEnv (because you're reparametrizing the sizes or something), but if you have two fake tensor modes which are sharing a shape environment, you would actually rather just reuse the original sizes/strides/offset from the original fake tensor. This ends up being pretty simple. I recommend viewing with whitespace diff turned off.

There's some fuzz around jagged tensor handling; that code is probably not quite right, but I fixed it for this particular case in the most straightforward way.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

voznesenskym committed:
Update on "If we re-fakeify a FakeTensor with the same ShapeEnv, preserve symbols"


Subsumes half of https://github.com/pytorch/pytorch/pull/113605

We support fakeifying an already fake tensor, which will give you a new fake tensor mirroring the same structure as the original fake tensor, which is what is needed by https://github.com/pytorch/pytorch/issues/113643 . However, when this refakeification happens, we will naively reallocate all new sizes for all of the fake tensor. This is the right thing to do if you are re-fakeifying on a fresh ShapeEnv (because you're reparametrizing the sizes or something), but if you have two fake tensor modes which are sharing a shape environment, you would actually rather just reuse the original sizes/strides/offset from the original fake tensor. This ends up being pretty simple. I recommend viewing with whitespace diff turned off.

There's some fuzz around jagged tensor handling; that code is probably not quite right, but I fixed it for this particular case in the most straightforward way.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
